### DISCUSSION TITLE
Lsp zero auto formatting

### DISCUSSION THREAD

**Discussion by IamTheLime** [VonHeikemen/lsp-zero.nvim#193] [Q&A]
I am a recent neovim adopter and this plugin has been a lifesaver! I have a question about the integration on the format_on_save feature shown in the docs.

I frequently program on python with a few legacy codebases (which unfortunately are not autoformatted) - for this I integrate null-ls using black and autopep8 with a plugin called lsp-format-modifications which ensures only git diffs are formatted. This works fine.

I have recently started taking up rust and I'd really like to get auto formatting setup, the issue I run into is that when I setup just the format_on_save, there is no action taken on save i.e. the file always remains the same 

```lua
lsp.format_on_save({
  servers = {
    ['lua_ls'] = {'lua'},
    ['rust_analyzer'] = {'rust'},
  }
})
```

The lsp.buffer_autoformat() on lsp attach will work every save but unfortunately breaks my aforementioned workflow.

I appreciate that this is a very niche situation, on a sepparate note I quite like the ability to chose the server used for the formatting, I was wondering if there is any particular reason for the behaviour of format_on_save. I will leave my config bellow (It is worth noting that I also tried it with a minimal lsp-zero config excluding all of the null-ls or format-modification nonsense).

```lua
local lsp = require('lsp-zero').preset({
  name = 'recommended',
  set_lsp_keymaps = true,
  manage_nvim_cmp = true,
  suggest_lsp_servers = true,
})

lsp.nvim_workspace()

lsp.on_attach(function(client, bufnr)
  vim.keymap.set({ 'n', 'x' }, 'gq', function()
    vim.lsp.buf.format({ async = false, timeout_ms = 10000 })
  end)
end)
lsp.format_on_save({
  servers = {
    ['lua_ls'] = {'lua'},
    ['rust_analyzer'] = {'rust'},
  }
})

-- This works, but forces every buffer to be formatted in full on save
-- lsp.on_attach(function(client, bufnr)
--  lsp.buffer_autoformat()
-- end)

lsp.setup()

vim.diagnostic.config({
  virtual_text = true,
  signs = true,
  update_in_insert = true,
  underline = false,
  severity_sort = false,
  float = true,
})



local null_ls = require("null-ls")
local null_opts = lsp.build_options('null-ls', {})
-- Note: I might want toremove this in the future
null_ls.setup({
  on_attach = function(client, bufnr)
    print(vim.inspect(client), bufnr)
    local lsp_format_modifications = require("lsp-format-modifications")
    lsp_format_modifications.attach(client, bufnr, { format_on_save = true })
    null_opts.on_attach(client, bufnr)
  end,
  sources = {
    null_ls.builtins.formatting.autopep8,
    null_ls.builtins.formatting.black,
  },
})

-- See mason-null-ls.nvim's documentation for more details:
-- https://github.com/jay-babu/mason-null-ls.nvim#setup
require('mason-null-ls').setup({
  ensure_installed = nil,
  automatic_installation = false, -- You can still set this to `true`
  automatic_setup = true,
})

-- Required when `automatic_setup` is true
require('mason-null-ls').setup_handlers()

```

**Comment by VonHeikemen** [comment]
That was probably a bug. I just fixed it, `format_on_save` should work now.

**Reply by IamTheLime** [reply]
Amazing, this fixed it!
